packages/transactions/lib/createMarketBuyTransaction.ts

```diff
@@ -0,0 +1,56 @@
+import _ from 'lodash'
+import { buy } from '@play-money/amms/lib/maniswap-v1'
+import { getMarket } from '@play-money/markets/lib/getMarket'
+import { TransactionItemInput, createTransaction } from './createTransaction'
+import { checkUserBalance } from './getUserBalances'
+
+interface MarketBuyTransactionInput {
+  userId: string
+  amount: number // in dollars
+  purchaseCurrencyCode: 'YES' | 'NO'
+  marketId: string
+}
+
+export async function createMarketBuyTransaction({
+  userId,
+  marketId,
+  amount,
+  purchaseCurrencyCode,
+}: MarketBuyTransactionInput) {
+  const market = await getMarket({ id: marketId })
+
+  const hasEnoughBalance = await checkUserBalance(userId, 'PRIMARY', amount)
+  if (!hasEnoughBalance) {
+    throw new Error('User does not have enough balance to make this purchase.')
+  }
+
+  let accumulatedTransactionItems: Array<TransactionItemInput> = []
+  let amountToPurchase = amount
+
+  while (amountToPurchase > 0) {
```

Comment:
I think the most correct way to do this is to have the exchanger run for the total amount before any of the AMMs/limit orders, and just return the shares to the buying account, and then separately the AMMs/limit orders take effect and the user gives the currency they don't want to the AMM/the other user in exchange for the currency they do want? And then you could have a squashing step to get rid of any transaction items that cancel each other out.

So, before the squash:
```
Gabrielle: -100 PRIMARY
Exchanger: +100 PRIMARY
Exchanger: -100 NO
Exchanger: -100 YES
Gabrielle: + 100 NO
Gabrielle: +100 YES
Gabrielle: -100 NO
AMM: +100 NO
AMM: -80 YES
Gabrielle: +80 YES
```

And after the squash:
```
Gabrielle: -100 PRIMARY
Exchanger: +100 PRIMARY
Exchanger: -100 NO
Exchanger: -100 YES
AMM: +100 NO
AMM: -80 YES
Gabrielle: +180 YES
```

Again though, this implementation also totally works, no need to change it.

Reply:
Oh this is a great idea. Do the exchange once with the user, and then you can hand out the shares to the AMM/limit orders as necessary. And then squash at the end!

Reply:
Refactored for this use case, much cleaner. Now just need to put in a squasher!